group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/assignor/StickyTaskAssignor.java

```diff
@@ -0,0 +1,401 @@
+                processIdToState.get(prevMember.processId).addTask(prevMember.memberId, task, true);
+                updateHelpers(prevMember, task, true);
+                it.remove();
+            }
+        }
+        // 2. re-assigning tasks to clients that previously have seen the same task (as standby task)
+        for (Iterator<TaskId> it = activeTasks.iterator(); it.hasNext();) {
+            final TaskId task = it.next();
+            final Set<Member> prevMembers = standbyTaskToPrevMember.get(task);
+            if (prevMembers != null && !prevMembers.isEmpty()) {
+                final Member prevMember = findMemberWithLeastLoad(prevMembers, task, true);
+                if (prevMember != null && hasUnfulfilledQuota(prevMember)) {
+                    processIdToState.get(prevMember.processId).addTask(prevMember.memberId, task, true);
+                    updateHelpers(prevMember, task, true);
+                    it.remove();
+                }
+            }
+        }
+        // 3. assign any remaining unassigned tasks
+        for (Iterator<TaskId> it = activeTasks.iterator(); it.hasNext();) {
+            final TaskId task = it.next();
+            final Member member = findMemberWithLeastLoad(task);
+            if (member != null) {
+                processIdToState.get(member.processId).addTask(member.memberId, task, true);
+                it.remove();
+                updateHelpers(member, task, true);
+            }
+        }
+    }
+
+    private void maybeUpdateTasksPerMember(final int activeTasksNo) {
+        if (activeTasksNo == tasksPerMember) {
+            totalCapacity--;
+            allTasks -= activeTasksNo;
+            tasksPerMember = computeTasksPerMember(allTasks, totalCapacity);
+        }
+    }
+
+    private Member findMemberWithLeastLoad(final Set<Member> members, TaskId taskId, final boolean returnSameMember) {
+        Set<Member> rightPairs = members.stream()
+            .filter(member  -> taskPairs.hasNewPair(taskId, processIdToState.get(member.processId).assignedTasks()))
+            .collect(Collectors.toSet());
+        if (rightPairs.isEmpty()) {
+            rightPairs = members;
+        }
+        Optional<ProcessState> processWithLeastLoad = rightPairs.stream()
+            .map(member  -> processIdToState.get(member.processId))
+            .min(Comparator.comparingDouble(ProcessState::load));
+        assert processWithLeastLoad.isPresent();
+        // if the same exact former member is needed
+        if (returnSameMember) {
+            return standbyTaskToPrevMember.get(taskId).stream()
+                .filter(standby -> standby.processId.equals(processWithLeastLoad.get().processId()))
+                .findFirst()
+                .orElseGet(() -> memberWithLeastLoad(processWithLeastLoad.get()));
+        }
+        return memberWithLeastLoad(processWithLeastLoad.get());
+    }
+
+    private Member findMemberWithLeastLoad(final TaskId taskId) {
```

Comment:
I think I get what's going on here, but I'm finding it a little challenging to follow exactly what the flow is with all these overloaded `findMemberWithLeastLoad` perhaps a short one-line comment for each?

Reply:
with some simplifications, now we only one `findMemberWithLeastLoad`